### PR TITLE
feat(qwen3-8-flash-next): add Engram training

### PR DESCRIPTION
**Pull request by HuiyingLi**
# What does this PR do ?

Add native Qwen3.8-Flash-Next 180B Engram training support with FlexAttention
sparse QSA, model-owned contiguous context parallelism, and reshardable Engram
checkpoint state.

# Changelog

- Add the Qwen3.8-Flash-Next model, config, registry entry, Hugging Face state-dict
  adapter, documentation, and 180B HellaSwag recipe.
- Add Engram N-gram Embedding ownership and checkpoint support, including DTensor-backed
  state for cross-topology DCP resharding.
- Execute sparse QSA through FlexAttention: route IDs scatter into a boolean
  membership table evaluated via a BlockMask, one code path for dense, packed
  (THD), and context-parallel layouts, retaining a bounded PyTorch oracle for
  CPU and parity.
- Add contiguous context parallelism for QSA, GatedDeltaNet, and Engram.
- Add focused unit tests for model construction, Engram, QSA, CP,
  FlexAttention, state-dict conversion, and checkpoint state.
- Add packed (THD) training: document-aware indexer routing with global
  flattened route IDs, GDN varlen pass-through, and loader `seq_lens` glue.
- Add packed context parallelism: local packed query shards attend to the
  gathered global K/V through the same FlexAttention path; packing is
  topology-independent for the model-owned CP path, so CP1 and CP-sharded
  runs consume byte-identical pack sequences.

### FlexAttention

- Kernel isolation (H100): CP-slice calls (`S_q=512`, gathered `KV=4096`) are
  bitwise identical to the full-row call; both within `0.18%` relative L2 of
  the fp32 gathered oracle.
- **50-step packed CP1 vs CP8 training parity** at world 64 / EP64 / GBS128
  over the same v3 Tulu-3 packs: loss-curve cosine `0.999999711`, mean
  absolute difference `0.00044`, maximum `0.0013` (4-decimal log
  quantization included), final losses `0.4916` (CP1) / `0.4912` (CP8).
  [CP1 run](https://wandb.ai/Nemo-automodel/bdelta/runs/bdelta-qwen4-tulu3-4k-cp1-world64-ep64-flex-packed-v1)
  / [CP8 run](https://wandb.ai/Nemo-automodel/bdelta/runs/bdelta-qwen4-tulu3-4k-cp8-world64-ep64-flex-packed-v1).

image

- Latest-head focused validation: `218 passed, 4 skipped` locally, covering the model suite plus affected checkpoint, MoE, training, and distributed tests.
- Ruff format/check passed.
- Diff checks passed.
- DCO check passed.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?

# Additional Information

- No related public issue.

### CODE REVIEW

**Review by akoumpa — DISMISSED** [review]

File: docs/model-coverage/llm/qwen/qwen3-8-flash-next.mdx [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+slug: model-coverage/large-language-models/qwen/qwen3-8-flash-next
+---
+
+Qwen3.8-Flash-Next is a hybrid-attention MoE model with four persistent HyperConnection
+streams and an Engram N-gram Embedding table (named `ple` in the checkpoint
+configuration). NeMo AutoModel
+supports the language-only `Qwen3_8_FlashNextForConditionalGeneration` path for
+supervised fine-tuning. All backbone parameters are trainable except the QSA
+indexers, whose reference path exposes discrete top-k IDs but no auxiliary
+loss or straight-through gradient.
```
Comment (jgerh) [inline docs/model-coverage/llm/qwen/qwen3-8-flash-next.mdx]:
```suggestion
Qwen3.8-Flash-Next is a hybrid-attention Mixture-of-Experts (MoE) model with four
persistent HyperConnection streams and an Engram N-gram Embedding table (named
`ple` in the checkpoint configuration). NeMo AutoModel supports the language-only
`Qwen3_8_FlashNextForConditionalGeneration` path for supervised fine-tuning
(SFT). All backbone parameters are trainable except the QSA indexers, whose
reference path exposes discrete top-k IDs but no auxiliary loss or
straight-through gradient.
```

File: docs/model-coverage/llm/qwen/qwen3-8-flash-next.mdx [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+- The Engram table has 320,001,536 padded rows of width 160. Its 51.2 billion
+  parameters are stored as 128 physical checkpoint shards.
+- The final learned HyperConnection read collapses four 2,560-wide streams to
+  one 2,560-wide state before the untied language-model head.
+
+## Validated Configuration
+
+| Area | Validated Configuration |
+|---|---|
+| FSDP2 | 47 decoder blocks use whole-layer activation checkpointing; the Engram owner-collective block remains eager; QSA indexers frozen |
```
Comment (jgerh) [inline docs/model-coverage/llm/qwen/qwen3-8-flash-next.mdx]:
```suggestion
| FSDP2 | 47 decoder blocks use whole-layer activation checkpointing; the Engram owner-collective block remains eager; QSA indexers are frozen |
```

File: docs/model-coverage/llm/qwen/qwen3-8-flash-next.mdx [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+99 and finished with validation loss 2.0610. The run completed without NaNs or
+collective errors and published model, optimizer, dataloader, scheduler, and
+RNG state in a distributed checkpoint.
+
+The Engram table is a globally shaped DTensor with placement `Shard(0)`. Request
+ranks group hashed row IDs by owner; an equal-capacity All-to-All sends IDs to
+owners and a second autograd-aware All-to-All returns values. Each rank directly
+looks up only its contiguous local DTensor shard, and backward reverses the
+value route so only the owner stores and updates each row's gradient. The
+validated cluster runs used NCCL's internal IB transport
```
Comment (jgerh) [inline docs/model-coverage/llm/qwen/qwen3-8-flash-next.mdx]:
```suggestion
The Engram table is a globally shaped DTensor with placement `Shard(0)`.
Requesting ranks group hashed row IDs by owner; an equal-capacity All-to-All
sends IDs to owners, and a second autograd-aware All-to-All returns values.
Each rank looks up only its contiguous local DTensor shard, and the backward
pass reverses the value route so only the owner stores and updates each row's
gradient. The validated cluster runs used NCCL's internal IB transport
```

File: docs/model-coverage/llm/qwen/qwen3-8-flash-next.mdx [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+(`NCCL_NET_PLUGIN=none` and `NCCL_NET=IB`).
+
+## Numerical Validation
+
+The released checkpoint was compared with its SGLang reference on two fixed
+raw-token sequences, including an EOS-reset case. Token embeddings, n-gram hash
```
Comment (jgerh) [inline docs/model-coverage/llm/qwen/qwen3-8-flash-next.mdx]:
```suggestion
raw-token sequences, including an EOS-reset case. Token embeddings, N-gram hash
```

File: docs/model-coverage/llm/qwen/qwen3-8-flash-next.mdx [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+IDs, and table values matched exactly. The first four decoder-layer states had
+cosine similarity from 0.9998 to 1.0000, and the corresponding logit-lens top-1
+token matched at every layer. The Engram delta had relative L2 error below 0.0063
+in both cases.
+
+The QSA routing path was also compared with SGLang on a fixed
+4,096-token input. Rows through position 2,050 matched the reference routing
+exactly. Replaying the exact SGLang layer input through AutoModel produced the
```
Comment (jgerh) [inline docs/model-coverage/llm/qwen/qwen3-8-flash-next.mdx]:
```suggestion
exactly. Replaying the exact SGLang layer input through NeMo AutoModel produced the
```

File: docs/model-coverage/llm/qwen/qwen3-8-flash-next.mdx [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+0.60801 for CP1 and 0.60800 for CP8.
+
+Packed (THD) context parallelism was validated the same way at world size 64
+with EP64: CP1 (DP64/GA2) and CP8 (DP8/GA16) trained 100 optimizer steps over
+identical THD packs (packing is topology-independent for the model-owned CP
+path, so both runs consume byte-identical pack sequences). Loss-curve cosine
+similarity was 0.999999925, mean absolute difference 0.00024, and maximum
+absolute difference 0.00069; final losses were 0.56053 for CP1 and 0.56011
+for CP8. A world-64 single-forward probe additionally matched CP8 against CP1
+with exact per-row routing IDs and passing Engram/GDN/QSA activation gates.
```
Comment (jgerh) [inline docs/model-coverage/llm/qwen/qwen3-8-flash-next.mdx]:
```suggestion
for CP8. A world-64 single-forward probe also matched CP8 against CP1 with exact
per-row routing IDs and passing Engram/GDN/QSA activation gates.
```

File: docs/model-coverage/llm/qwen/qwen3-8-flash-next.mdx [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+
+## Example Recipe
+
+| Recipe | Description |
+|---|---|
+| [qwen3_8_flash_next_180b_hellaswag_ep64.yaml](https://github.com/NVIDIA-NeMo/Automodel/blob/main/examples/llm_finetune/qwen/qwen3_8_flash_next_180b_hellaswag_ep64.yaml) | 100-step SFT on HellaSwag with FSDP2, EP64, and world-sized Engram ownership |
+
+The recipe loads [Qwen/Qwen3.8-Flash-Next](https://huggingface.co/Qwen/Qwen3.8-Flash-Next)
+directly from the Hugging Face Hub. Allocate eight nodes with eight GPUs per
```
Comment (jgerh) [inline docs/model-coverage/llm/qwen/qwen3-8-flash-next.mdx]:
```suggestion
directly from the Hugging Face Hub. Allocate 8 nodes with 8 GPUs per node and
launch the recipe through your cluster launcher.
```

File: docs/model-coverage/llm/qwen/qwen3-8-flash-next.mdx [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+- Native distributed checkpoints store the Engram parameter and optimizer tensors
+  under their global DTensor FQNs, so DCP can reshard them when the owner world
+  size changes. Checkpoints from the earlier rank-local owner format are not
+  supported; resume from a native DTensor checkpoint or restart the optimizer.
+- CP is limited to language-only inputs with TP=PP=1. Non-packed
+  right-tail-padded batches and packed (THD) batches with ``cu_seqlens`` (or
+  loader ``seq_lens``) are both supported; the same FlexAttention path serves
+  packed CUDA training and packing is performed topology-independently.
+  Left/interior padding, TP, PP, MTP, and vision/video training remain
+  unsupported.
```
Comment (jgerh) [inline docs/model-coverage/llm/qwen/qwen3-8-flash-next.mdx]:
```suggestion
  right-tail-padded batches and packed (THD) batches with `cu_seqlens` (or
  loader `seq_lens`) are both supported; the same FlexAttention path serves
  packed CUDA training, and packing is performed topology-independently.
  Left padding, interior padding, TP, PP, MTP, and vision and video training
  remain unsupported.
```

**Review by jgerh — COMMENTED** [review]
Completed tech pubs review of docs/model-coverage/llm/qwen/qwen3-8-flash-next.mdx and provided a few copyedits.

File: nemo_automodel/components/models/qwen3_8_flash_next/flex_qsa.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        query.permute(0, 2, 1, 3),
+        key.permute(0, 2, 1, 3),
+        value.permute(0, 2, 1, 3),
+        block_mask=block_mask,
+        scale=scale,
+        enable_gqa=True,
+    ).permute(0, 2, 1, 3)
+    # Padding-query rows (no routes) must be exactly zero in both output and
+    # gradient, matching the oracle and the padded-gap training contract.
+    return output * has_routes[:, :, None, None].to(output.dtype)
```
[the remark this answers is not in the document]
Comment (HuiyingLi) [reply nemo_automodel/components/models/qwen3_8_flash_next/flex_qsa.py]: Fixed in 229067416: route-empty rows now get a dummy K/V-0 membership before FlexAttention, and masked_fill erases their output afterward. Added a CPU membership invariant plus a real H100 CUDA forward/backward test; the padding-row output and Q/K/V gradients are exactly zero and finite.

**Review by akoumpa — APPROVED** [review]
